src/shared_utilities/SpaceVaryingInputs.jl

```diff
@@ -0,0 +1,169 @@
+of z in the model grid - defined implicitly by `space`.
+
+Returns a ClimaCore.Fields.Field of scalars.
+"""
+function SpaceVaryingInput(
+    data_z::AbstractArray,
+    data_values::AbstractArray,
+    space::S,
+) where {S <: ClimaCore.Spaces.CenterFiniteDifferenceSpace}
+    model_value = ClimaCore.Fields.zeros(space)
+    # convert the passed arrays to the appropriate type for the device
+    device = ClimaComms.device(space)
+    AT = ClimaComms.array_type(device)
+    data_values = AT(data_values)
+    data_z = AT(data_z)
+    zvalues = ClimaCore.Fields.coordinate_field(space).z
+    #now create the parameter field
+    model_value .= map(zvalues) do z
+        linear_interpolation(data_z, data_values, z)
+    end
+    return model_value
+end
+
+
+"""
+    SpaceVaryingInputs.SpaceVaryingInput(
+        data_z::AbstractArray,
+        data_values::NamedTuple,
+        space::S,
+        dest_type::Type{DT},
+    ) where {
+        S <: ClimaCore.Spaces.CenterFiniteDifferenceSpace,
+        DT,
+    }
+
+Returns a field of parameter structs to be used in the model; 
+appropriate when the parameter struct values vary in depth; 
+the `dest_type` argument is the struct type - we assumed that
+your struct as a constructor which accepts the values of its arguments
+by kwarg,
+- `data_z` is where the measured values were obtained,
+- `data_values` is a NamedTuple with keys equal to the argument names
+of the struct, and with values equal to an array of measured values,
+- `space` defines the model grid.
+
+As an example, we can create a field of vanGenuchten structs as follows. This struct
+requires two parameters, `α` and `n`. Let's assume that we have measurements of these
+as a function of depth at the locations given by `data_z`, called `data_α` and `data_n`.
+Then we can write
+`vG_field = SpaceVaryingInput(data_z, (;α = data_α, n = data_n), space, vanGenuchten{Float32})`.
+Under the hood, at each point in the model grid, we will create
+`vanGenuchten{Float32}(;α = interp_α, n = interp_n)`, where `interp` indicates
+the interpolated value at the model depth.
+
+Returns a ClimaCore.Fields.Field of type DT.
+
+"""
+function SpaceVaryingInput(
+    data_z::AbstractArray,
+    data_values::NamedTuple,
+    space::S,
+    dest_type::Type{DT},
+) where {S <: ClimaCore.Spaces.CenterFiniteDifferenceSpace, DT}
+    zvalues = ClimaCore.Fields.coordinate_field(space).z
+    # convert the passed arrays to the appropriate type for the device
+    device = ClimaComms.device(space)
+    AT = ClimaComms.array_type(device)
+    data_z = AT(data_z)
+    #=
+    data_values_AT = map(keys(data_values)) do key
+        key => AT(getproperty(data_values, key))
+    end
+    data_values_AT = NamedTuple(data_values_AT)
+    =#
+    # now create the field of structs
+    model_value = map(zvalues) do z
+        args = map(data_values) do AT(value)
+            return linear_interpolation(data_z, value, z)
+        end
+        DT(; args...)
```

Comment:
Okay, now the issue is that `DT` is not isbits.

Reply:
@dennisYatunin can you help us here?

What we are trying to do here is to create a Field of struct where the type of struct is dynamically determined. This function `SpaceVaryingInput` will be used to set up general space-varying parameter fields.

Reply:
Oh actually this just looks like invalid syntax. You can't write `do AT(value)`; you need to write `do value` and then pass `AT(value)` to the `linear_interpolation`.

Reply:
DT should be ok, I think. Just a struct of floats

we are converting the depth coordinates (data_z) to AT, and then also each value of a NamedTuple of Vectors. let me try what you suggested, and happy to refactor too

Reply:
I dont think we can call AT(value) within the loop because AT itself is not isbits

Reply:
yay thank you @dennisYatunin !